crates/rslint_parser/src/ast.rs

```
@@ -185,9 +185,163 @@ impl<N: AstNode> IntoIterator for AstNodeList<N> {
+		}
+	}
+
+	/// Returns an iterator over all separator tokens
+	pub fn separators(&self) -> impl Iterator<Item = SyntaxToken> {
+		self.elements()
+			.flat_map(|element| element.trailing_separator)
+	}
+
+	/// Returns an iterator over all nodes
+	pub fn iter(&self) -> AstSeparatedListNodesIterator<N> {
+		AstSeparatedListNodesIterator {
+			inner: self.elements(),
+		}
+	}
+
+	#[inline]
+	pub fn is_empty(&self) -> bool {
+		self.len() == 0
+	}
+
+	pub fn len(&self) -> usize {
+		// TODO 1724 replace with (self.list.len() + 1) << 2 once trivia are attached to tokens
+		self.iter().count()
+	}
+
+	pub fn trailing_separator(&self) -> Option<SyntaxToken> {
+		// TODO 1724: Replace with simple match once trivia is no longer stored in the body
+		// match self.list.last() {
+		// 	NodeOrToken::Token(token) => Some(token),
+		// 	_ => None
+		// }
+
+		self.elements().last()?.trailing_separator
+	}
+}
+
+#[derive(Debug, Clone)]
+pub struct AstSeparatedListElementsIterator<N> {
+	next: SyntaxElementChildren,
+	ph: PhantomData<N>,
+}
+
+impl<N> AstSeparatedListElementsIterator<N> {
+	// TODO 1724: Replace with call to next once trivia are no longer stored in tokens and errors are part of the union types.
+	fn next_non_trivia_or_error(&mut self) -> Option<SyntaxElement> {
+		self.next.find_map(|element| match &element {
+			NodeOrToken::Node(node) => {
+				if node.kind() == SyntaxKind::ERROR {
+					None
+				} else {
+					Some(element)
+				}
+			}
+			NodeOrToken::Token(token) => {
+				if token.kind().is_trivia() {
+					None
+				} else {
+					Some(element)
+				}
+			}
+		})
+	}
+}
+
+impl<N: AstNode> Iterator for AstSeparatedListElementsIterator<N> {
+	type Item = AstSeparatedElement<N>;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let node_or_token = self.next_non_trivia_or_error()?;
+
+		let element = match node_or_token {
+			// The node for this element is missing if there's right a separator token
+			NodeOrToken::Token(token) => panic!(
```

Is using `panic!` the right choice? If so, should we leave a comment? 
I would expect a diagnostic of something else to handle gracefully the exception.

---

This is one of the consequences of the new AST facade. Our new AST facade gives new guarantees that will require panics/asserts to satisfy Rust's type checker. For example, we define that the 3rd element of an `if` statement will always be the condition. That means, we'll need to *cast* to a condition and *unwrap* the node. 

The parser should never generate a parse tree that validates this assumption. It's a parser bug if it does. 

That's why working on the syntax level is highly discouraged and our mutation API should ideally only work on the AST level, so that invalidating the tree never is possible (and, therefore, casting is safe).

---

> That's why working on the syntax level is highly discouraged and our mutation API should ideally only work on the AST level, so that invalidating the tree never is possible (and, therefore, casting is safe).

This is something we may need to put more thought into. We probably want something like Swift's make/with/builder APIs that work at the AST level, and that should be the preferred way of working with syntax nodes on a small scale.

But I think it's very useful to have some methods like those in Roslyn's [SyntaxNodeExtensions](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxnodeextensions?view=roslyn-dotnet-3.11.0) (ReplaceNodes especially) and I'm not sure where exactly we fit tree validation into that.

I think we should try to avoid panics even if transforms misbehave, so if we need to panic on certain invalid states, we'd want to prevent transforms from being able to create those states. So that would mean going a step further than just discouraging certain APIs and we should consider which ones we'd have to rule out entirely if we can't otherwise avoid these sorts of panics.

---

> This is something we may need to put more thought into. We probably want something like Swift's make/with/builder APIs that work at the AST level, and that should be the preferred way of working with syntax nodes on a small scale

Certainly, but that's something I consider out of scope for the parser work. Also if it is indeed limited to the small scale. 


> But I think it's very useful to have some methods like those in Roslyn's SyntaxNodeExtensions (ReplaceNodes especially) and I'm not sure where exactly we fit tree validation into that.

These may be useful but I still think that most refactorings/lint autofixes should have no need for these extensions. The main use case that I see for working with `SyntaxNode` directly is if you want to implement something generically, regardless of the node type. 

> I think we should try to avoid panics even if transforms misbehave, so if we need to panic on certain invalid states, we'd want to prevent transforms from being able to create those states. So that would mean going a step further than just discouraging certain APIs and we should consider which ones we'd have to rule out entirely if we can't otherwise avoid these sorts of panics.

I'm not sure if we can make it impossible but agree, we should make it as hard as possible for transforms to generate invalid source code. 

Casting here just reveals an error in the tree after a transform (or parser bug) that would result in a syntax error when emitting the file. Adding checks to avoid emitting files with syntax errors will also prevent that the cast here can fail.